src/utils.py

```diff
@@ -1,45 +1,36 @@
-import src.config as config
 import os
-import pandas as pd
+from dateutil import parser
+import pyarrow.parquet as pq
+import pyarrow as pa
+from src.data_scraper import time_helpers
 
 
-def save(data, subfolder, filename, overwrite=False):
-    """WIP. Save the loaded data. If a file already exists, it will try to append to it by columns or indexes.
-    If duplicates exist, it will only preserve the last, remove the previous ones, and save the file again."""
-    os.makedirs(os.path.join(config.FOLDER_TO_SAVE, subfolder), exist_ok=True)
-    if overwrite:
-        data.to_csv(os.path.join(config.FOLDER_TO_SAVE, subfolder, filename), index=False)
-    else:
-        if filename in os.listdir(os.path.join(config.FOLDER_TO_SAVE, subfolder)):
-            saved_data = pd.read_csv(os.path.join(config.FOLDER_TO_SAVE, subfolder, filename))
-            data = data.astype(object)
-            saved_data = saved_data.astype(object)
-            if sorted(data.columns.tolist()) == sorted(saved_data.columns.tolist()):
-                data = pd.merge(data, saved_data, how='outer', on=data.columns.tolist())
-            else:
-                # Find common columns to merge on
-                common_cols = data.columns[data.columns.isin(saved_data.columns)].tolist()
-                data = data.merge(saved_data, how='outer', on=common_cols, copy=False)
-
-        data = data.groupby(config.MERGE_DATA_ON).last().reset_index()
-        data = data.sort_values(by=config.MERGE_DATA_ON)
-        data.to_csv(os.path.join(config.FOLDER_TO_SAVE, subfolder, filename), index=False)
+def dataset_stored(dataset_path, start_time, end_time) -> bool:
+    """
+    Checks availability of stored data on disk for selected start and end dates.
+    The data is available if requested start date >= saved start date and requested end date >= saved end date.
+    """
+    if os.path.exists(dataset_path):
+        available_dates = os.listdir(dataset_path)
+        available_dates = [date.split('date=')[-1] for date in available_dates if not date.startswith('.')]
+        available_dates = [parser.parse(date).date() for date in available_dates]
 
+        start_date = time_helpers.timestamp_to_datetime(start_time).date()
+        end_date = time_helpers.timestamp_to_datetime(end_time).date()
 
-def load_from_checkpoint(subfolder, filename, start_time, end_time):
-    """WIP"""
-    if filename in os.listdir(os.path.join(config.FOLDER_TO_SAVE, subfolder)):
-        test = pd.read_csv(f'{config.FOLDER_TO_SAVE}/{subfolder}/{filename}')
-        # If the last timestamp in the saved file > specified end timestamp
-        if test['exact_time'].astype('datetime64[ns]').iloc[-1] > end_time:
-            if test.iloc[0]['exact_time'] < start_time:
-                print('File already saved.')
-            else:
-                end_time = int(test.iloc[0]['exact_time'])
+        return start_date >= min(available_dates) and end_date <= max(available_dates)
```

Comment:
What if:

1. We have two datasets stored from 2018 till 2019 and from 2020 till 2021.
2. We ask to retrieve from 2018 till 2021.
Wouldn't it think that it has all the available data?

Reply:
Yes, it would fail to recognize it. It's not perfect at all but I feel like we should prioritize speed over quality now. So I'd rather move on with writing the backtesting module instead and leave solving these cases for later. :)